fiftyone/utils/eval/classification.py

```diff
@@ -0,0 +1,499 @@
+                }
+            ),
+        ).save(eval_field)
+
+    # Equivalent with loops
+    # ytrue = []
+    # ypred = []
+    # confs = []
+    # with fou.ProgressBar() as pb:
+    #     for sample in pb(samples.select_fields([pred_field, gt_field])):
+    #         gt_label = sample[gt_field].label
+    #         pred_label = sample[pred_field].label
+    #         pred_conf = sample[pred_field].confidence
+    #         ytrue.append(gt_label)
+    #         ypred.append(pred_label)
+    #         confs.append(pred_conf)
+    #         if eval_field:
+    #             if gt_label == pos_label:
+    #                 eval_label = "TP" if pred_label == pos_label else "FN"
+    #             else:
+    #                 eval_label = "TN" if pred_label != pos_label else "FP"
+    #
+    #             sample[eval_field] = fol.Classification(label=eval_label)
+    #             sample.save()
+
+    return BinaryClassificationResults(ytrue, ypred, confs, classes)
+
+
+def evaluate_top_k_classifications(
+    samples, k, classes, pred_field, gt_field="ground_truth", eval_field=None,
+):
+    """Evaluates the top-k accuracy of the classification predictions in the
+    given samples with respect to the specified ground truth labels.
+
+    The predictions in ``pred_field`` must have their ``logits`` populated.
+
+    If an ``eval_field`` is specified, this method will record whether each
+    prediction is top-k correct in this field.
+
+    Args:
+        samples: a :class:`fiftyone.core.collections.SampleCollection`
+        k: the top-k value to use when assessing accuracy
+        classes: the list of class labels corresponding to the predicted logits
+        pred_field: the name of the field containing the predicted
+            :class:`fiftyone.core.labels.Classification` instances to evaluate.
+            This field must have its ``logits`` populated
+        gt_field ("ground_truth"): the name of the field containing the ground
+            truth :class:`fiftyone.core.labels.Classification` instances
+        eval_field (None): the name of a field in which to record whether each
+            prediction is top-k correct
+
+    Returns:
+        the top-k accuracy in ``[0, 1]``
+    """
+    targets_map = {label: idx for idx, label in enumerate(classes)}
+
+    # This extracts a `num_samples x num_classes` array of logits
+    ytrue, logits = samples.aggregate(
+        [foa.Values(gt_field + ".label"), foa.Values(pred_field + ".logits")]
+    )
+
+    correct = []
+    for _label, _logits in zip(ytrue, logits):
+        if _logits is not None:
+            target = targets_map[_label]
+            top_k = np.argpartition(_logits, -k)[-k:]
+            _correct = target in top_k
+        else:
+            _correct = False
+
+        correct.append(_correct)
+
+    if eval_field:
+        samples._add_field_if_necessary(eval_field, fof.BooleanField)
+        samples.set_values(eval_field, correct)
```

Comment:
This one is too complex to be done within an aggregation, so we use `values()` to extract all the values from the DB and then `set_values()` to write the output in bulk.

This is *significantly* faster than the equivalent loop I commented out below.